tests/non-interference.html

```diff
@@ -0,0 +1,211 @@
+
+    assert_equals(decoder.decode(new Uint8Array([65])), 'A',
+                  'decode should work');
+
+    writer.write(new Uint8Array(bytes.slice(3)));
+
+    return reader.read();
+  }).then(({value, done}) => {
+    assert_false(done, 'should not be done');
+    assert_equals(value, string.substring(2),
+                  'split character should have been assembled');
+  });
+}, 'decode() should not be influenced by transform state');
+
+// This is the same test, but with the roles reversed.
+promise_test(() => {
+  const decoder = new TextDecoder();
+
+  assert_equals(decoder.decode(new Uint8Array(bytes.slice(0, 3)),
+                               { stream: true }),
+                string.substring(0, 2),
+                'result should not include partial characters');
+
+  const writer = decoder.writable.getWriter();
+  writer.write(new Uint8Array([65]));
+
+  const reader = decoder.readable.getReader();
+  return reader.read().then(({value, done}) => {
+    assert_false(done, 'should not be done');
+    assert_equals(value, 'A', 'transform should work');
+
+    assert_equals(decoder.decode(new Uint8Array(bytes.slice(3))),
+                  string.substring(2),
+                  'split character should have been assembled');
+  });
+}, 'transform should not be influenced by decode() state');
+
+promise_test(() => {
+  const decoder = new TextDecoder();
+  const writer = decoder.writable.getWriter();
+
+  writer.write(new Uint8Array([65]));
+  writer.close();
+
+  const reader = decoder.readable.getReader();
+  return reader.read().then(({value, done}) => {
+    assert_false(done, 'should not be done');
+    assert_equals(value, 'A', 'transform should have worked');
+
+    return reader.read();
+  }).then(({value, done}) => {
+    assert_true(done, 'should be done');
+
+    return Promise.all([reader.closed, writer.closed]);
+  }).then(() => {
+    assert_equals(decoder.decode(bytesAsArray), string,
+                  'decode should still work');
+  });
+}, 'decode() should still work after transform is closed');
+
+promise_test(() => {
+  const decoder = new TextDecoder();
+  const writer = decoder.writable.getWriter();
+
+  return writer.abort().then(() => {
+    assert_equals(decoder.decode(bytesAsArray), string,
+                  'decode should still work');
+  });
+}, 'decode() should still work after transform is aborted');
+
+promise_test(t => {
+  const decoder = new TextDecoder('utf-8', { fatal: true });
+  const writer = decoder.writable.getWriter();
+
+  // 0xFF is never a valid byte in UTF-8.
+  writer.write(new Uint8Array([0xFF]));
+
+  const reader = decoder.readable.getReader();
+  return promise_rejects(t, new TypeError(), reader.read(),
+                         'read() should throw').then(() => {
+    assert_equals(decoder.decode(bytesAsArray), string,
+                  'decode should still work');
+  });
+}, 'decode() should still work after transform is errored');
```

Comment:
Should there be 'encode() should still work after transform is errored' for symmetry?

Reply:
It's a little bit different, since the UTF-8 encoding process itself will never error the stream. But converting the chunk to a string can error the stream, so I added a test that does that.